### PR TITLE
Add 24bit packed formats to ALSA, dummy

### PR DESCRIPTION
**Pull request by 0joshuaolson1**

### CODE REVIEW

**Comment by andrewrk** [comment]
may I propose these names instead?

 * `SoundIoFormatS24PackedNE`
 * `SoundIoFormatS24PackedFE`
 * `SoundIoFormatS24PackedLE`
 * `SoundIoFormatS24PackedBE`
 * `SoundIoFormatU24PackedNE`
 * `SoundIoFormatU24PackedFE`
 * `SoundIoFormatU24PackedLE`
 * `SoundIoFormatU24PackedBE`

File: src/soundio.c [hunk 1]
```diff
@@ -111,6 +111,10 @@ int soundio_get_bytes_per_sample(enum SoundIoFormat format) {
     case SoundIoFormatS32BE:      return 4;
     case SoundIoFormatU32LE:      return 4;
     case SoundIoFormatU32BE:      return 4;
+    case SoundIoFormatS243LE:     return 1;////////?
```
Comment (andrewrk) [inline src/soundio.c]: 3 for all of them.

File: src/soundio.c [hunk 2]
```diff
@@ -137,6 +141,10 @@ const char * soundio_format_string(enum SoundIoFormat format) {
     case SoundIoFormatS32BE:      return "signed 32-bit BE";
     case SoundIoFormatU32LE:      return "unsigned 32-bit LE";
     case SoundIoFormatU32BE:      return "unsigned 32-bit BE";
+    case SoundIoFormatS243LE:     return "signed 24-bit 3bytes LE";
```
Comment (andrewrk) [inline src/soundio.c]: How about "signed 24-bit packed LE" ?

**Comment by andrewrk** [comment]
Looks good. After fixing `soundio_get_bytes_per_sample` are you able to test it with your DAC?

**Comment by 0joshuaolson1** [comment]
I'll test in the next couple days and make the changes. I have some coursework to take care of.

**Comment by 0joshuaolson1** [comment]
It works! Since my output device is stereo and S24PackedLE:, it seems that the pointer to the SoundIoChannelArea struct (let's call it `frames`) can be used like this:
```
frames[chan].ptr[0] = 0; // chan is 0 or 1, zero assignments aren't necessary
frames[chan].ptr[1] = 0; // cast to a char(?), and the middle significant byte
frames[chan].ptr[2] = 0; // most significant byte of signed 24-bit integer
frames[chan].ptr += frames[chan].step;
```
I can make square/pulse waves, so I verified the endian behavior by loudness according to which byte

What I don't understand is why frames[chan].step is 6. Am I skipping over assigning some frames by doing pointer arithmetic this way?

**Comment by andrewrk** [comment]
You shouldn't be using [0] [1] [2] on ptr, instead rely on `+= step` to get to the next frame. See the sio_sine for example.

This abstracts over the fact that some streams are planar and some are interleaved, and will make your code work for both.

**Comment by 0joshuaolson1** [comment]
If we're using the same definition of frame, I am using step. Every `write_sample` in sio_sine writes a value directly to the pointer, but there is no 24-bit machine type. Should I be casting each assignment to chars? Why is there a step for each channel if they could be interleaved?

I'll see what I can do with PulseAudio.

**Comment by andrewrk** [comment]
Ah I see, my mistake. If the step value is 6 that means the stream is interleaved, so likely `frames[other_chan].ptr == frames[chan].ptr + 3`.

**Comment by 0joshuaolson1** [comment]
Instead opening one with PulseAudio too.

**Pull request closed without merging**